src/dosext/misc/emm.c

```diff
@@ -273,33 +273,44 @@ static void set_map_registers(const struct emm_reg *buf, int pages);
 #define Kdebug1(args)		E_Stub args
 #define Kdebug2(args)		E_Stub args
 
+static int ems_init_done = 0;
+
 void
 ems_helper(void) {
   switch (LWORD(ebx)) {
   case 0:
     E_printf("EMS Init called!\n");
     if (!config.ems_size) {
+      LWORD(eax) = 1;	/* report error */
       LWORD(ebx) = EMS_ERROR_DISABLED_IN_CONFIG;
       return;
     }
+    if (ems_init_done) {
+      LWORD(eax) = 1;	/* report error */
+      LWORD(ebx) = EMS_ERROR_ALREADY_INITIALISED;
+      return;
+    }
     if (HI(ax) < DOSEMU_EMS_DRIVER_MIN_VERSION) {
       error("EMS driver version mismatch: got %i, expected %i, disabling.\n"
             "Please update your ems.sys from the latest dosemu package.\n",
         HI(ax), DOSEMU_EMS_DRIVER_VERSION);
       com_printf("\nEMS driver version mismatch, disabling.\n"
             "Please update your ems.sys from the latest dosemu package.\n"
             "\nPress any key!\n");
+      LWORD(eax) = 1;	/* report error */
       LWORD(ebx) = EMS_ERROR_VERSION_MISMATCH;
       set_IF();
       com_biosgetch();
       clear_IF();
       return;
     }
+
     if (HI(ax) < DOSEMU_EMS_DRIVER_VERSION) {
       warn("EMS driver too old, consider updating %i->%i\n",
         HI(ax), DOSEMU_EMS_DRIVER_VERSION);
       com_printf("EMS driver too old, consider updating.\n");
     }
+    ems_init_done = 1;	/* make sure the driver isn't loaded twice */
```

Comment:
>I think it would be better to do
the EMS check in ems.sys itself.
If EMS is present, it should not
load. Its because if you use the
native himem.sys and try to load
it twice, it gracefully fails without
any help from dosemu.

But what would you use to check in ems.sys, there's no persistent memory between attempted loads of the device driver, unless you want to test the interrupt vector or the sniff the device driver chain for name 'EMMXXXX0'?

Assuming we find another EMS driver already installed somehow, does that mean the XMS part of this driver shouldn't install either?